rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,350 @@
+
+#### 5. Start EC2 Instance to Download File
+
+Spin up EC2 instances with enough storage to download the whole file. TODO: See if spot instances could be used.
+
+#### 6. Download the File
+
+The most time will be spent here, downloading the file. Most CSP offer a way to validate the integrity of your download 
+by providing a hash of the data. We can compare this hash with the hash of downloaded file. If the hash does not match 
+we can reattempt the download.
+
+#### 7. Upload to S3
+
+If the file is >5GB in size we can upload the file directly to s3 without writing to disk. For files <5GB, the file must
+be completely downloaded before it can be uploaded to s3. The AWS SDK will be used to upload large files in parts.
+Once upload is complete, and the download job will be removed from the queue. 
+
+Before uploading we will need to calculate
+the expected hash. We can use the [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax) 
+ during a put request to verify the integrity of each chunk.
+ 
+> ??? Do we need to tell the the user or the DB when the file is done do uploading. This makes me think we need an endpoint
+> to finalize the upload. The endpoint will look for the expected file, and if it's found, update the database with the file
+> info.
+
+### Local to Cloud Flow
+
+Local to cloud upload will allow users to upload files from their local machine to the cloud. Uploads happen though the
+DP browser App directly to S3. For small files <5GB, uploads can be accomplished using presigned URLS. For large files >5GB
+the user must perform a multipart upload. Now to walk though the local to cloud flow, see figure 2.
+
+![Local to Cloud](https://app.lucidchart.com/publicSegments/view/80361b71-5317-478b-921e-0ead5152c865/image.png)
+**Figure 3:** Architecture of a user uploading from their local machine to our cloud.
+
+#### 1. Upload File
+
+Using the browser App, the user select a file to upload and send a request to upload the file.
+
+#### 2. Request Token or Presigned URL
+
+Depending on the size of the file a different response will be returned. If the file is >5GB
+
+#### 3. Upload Directly S3
+
+Using the access token and the AWS JS SDK we will upload the file directly to S3. We will calculate the hash of each chunk
+locally and add the hash to [Content-MD5 header](https://docs.aws.amazon.com/AmazonS3/latest/API/API_PutObject.html#API_PutObject_RequestSyntax) 
+in the Put Object request. This will verify the integrity of each chunk during the upload. If the hash does not match
+the chunk will be rejected.
+
+##### User Upload Limitation
+
+S3 is our CSP, and is where a user's uploaded files will eventually end up. S3 will allow small files to be upload without
+using multipart upload. This allows us to generate a single presigned URL that the user can use to securely upload their
+files.
+
```

Comment:
Consider _S3 is the CZI CSP where uploaded datasets are stored. Since S3 allows small files to be uploaded without using multipart upload, our implementation can generate a single presigned URL when securely uploading small datasets._